Sources/System/FileOperations.swift

```diff
@@ -370,4 +370,26 @@ extension FileDescriptor {
   public func dup2() throws -> FileDescriptor {
     fatalError("Not implemented")
   }
+  
+  /// Create a pipe, a unidirectional data channel which can be used for interprocess communication.
+  ///
+  /// - Returns: The pair of file descriptors.
+  ///
+  /// The corresponding C function is `pipe`.
+  @_alwaysEmitIntoClient
+  // @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+  public static func pipe() throws -> (readEnd: FileDescriptor, writeEnd: FileDescriptor) {
+    try _pipe().get()
+  }
+  
+  @usableFromInline
+  internal static func _pipe() -> Result<(readEnd: FileDescriptor, writeEnd: FileDescriptor), Errno> {
+    var fds: (Int32, Int32) = (-1, -1)
+    return withUnsafeMutableBytes(of: &fds) { bytes in
+      let fds = bytes.bindMemory(to: Int32.self)
+      return valueOrErrno(retryOnInterrupt: false) {
+        system_pipe(fds.baseAddress!)
+      }.map { _ in (FileDescriptor(rawValue: fds[0]), FileDescriptor(rawValue: fds[1])) }
+    }
```

Comment:
You might be able to say `var fds: Array<CInt> = [-1, -1] ` and use the implicit array-to-pointer conversion (thought I don't recall if it's mutable). Otherwise, you can probably use `withUnsafeMutablePointer` to avoid the extra bind-memory step.

Reply:
@lorentey suggested I use the tuple form for a stronger guarantee that those values would end up on the stack. I haven't checked but I believe `withUnsafeMutablePointer` will still require rebinding the type (I think this was a signedness issue), but would also require manually passing the count (1), whereas `withUnsafeMutableBytes` takes this information from the buffer pointer.

Reply:
Please do make sure that this builds on windows - I'm almost certain that this will break the Windows builds.  Also note that Windows should transact in `HANDLE`s, aka `void *`, so this is going to truncate at least.  See [`CreatePipe`](https://docs.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-createpipe).

Reply:
I believe it's fine to simply surround these additions with `#if !os(Windows)` for now.

Reply:
This is looking great! @atrick / @glessard , what's the lesser of evils for the pointer binding stuff?

Reply:
The pointer to a homogenous tuple is also bound to its element type, so you can use `assumingMemoryBound`.

_(I hope. I use this quite a lot)_

Reply:
As written, this case indeed calls for `assumingMemoryBound`.  
The memory is _already_ bound to `Int32`, so `bindMemory` doesn't add information.

This being said, we unfortunately don't have `assumingMemoryBound` on the `RawBufferPointer` types at this point. We also gain nothing from using the Buffers in the case: we must rely in the C call to be well behaved instead of having any sort of bounds checking. Given that, I suggest this for lines 389-393:
```
    return withUnsafeMutablePointer(to: &fds) { pointer in
      valueOrErrno(retryOnInterrupt: false) {
        system_pipe(UnsafeMutableRawPointer(pointer).assumingMemoryBound(to: Int32.self))
      }
    }.map { _ in (FileDescriptor(rawValue: fds.0), FileDescriptor(rawValue: fds.1)) }
```

Reply:
@karl is correct, and @glessard 's code edit looks good.
Although the original code was harmless, it doesn't really make sense to bind memory that belongs to a variable.

Reply:
Is there a document detailing what "binding memory" actually does? I had thought this was just managing compile-time information so I'm not 100% clear on the difference between "assuming" memory bound and actually binding memory.

Reply:
There is some scattered documentation, including the doc-comments, but the largest chunk is in the `RawPointer` proposal: https://github.com/apple/swift-evolution/blob/main/proposals/0107-unsaferawpointer.md  
Experience has shown that those sources are insufficient.

@NevinBR (I think) came up with a nice description of binding for humans a few weeks ago: https://forums.swift.org/t/pitch-implicit-pointer-conversion-for-c-interoperability/51129/36.

In general, if you're reminding the compiler of type information it should already have known, but had been obscured for some reason, then `assumingMemoryBound` is the thing to reach for (as we did here). If you are telling the compiler new information, then reach for `bindMemory` or `withMemoryRebound`.